cpp/src/Ice/Instance.cpp

```diff
@@ -1283,6 +1301,33 @@ IceInternal::Instance::initialize(const Ice::CommunicatorPtr& communicator)
     }
 }
 
+const Ice::ImplicitContextPtr&
+IceInternal::Instance::getImplicitContext() const
+{
+    switch (_implicitContextKind)
+    {
+        case ImplicitContextKind::PerThread:
+        {
+            static thread_local std::map<const IceInternal::Instance*, ImplicitContextPtr> perThreadImplicitContextMap;
+            auto it = perThreadImplicitContextMap.find(this);
+            if (it == perThreadImplicitContextMap.end())
+            {
+                auto r = perThreadImplicitContextMap.emplace(make_pair(this, std::make_shared<ImplicitContext>()));
+                return r.first->second;
+            }
+            else
+            {
+                return it->second;
+            }
+        }
+
+        default:
+        {
+            return _sharedImplicitContext;
```

Comment:
When ImplicitContextKind::None, we return a dereferenced null shared_ptr?

Reply:
I assume the caller ensures this never happens, so I would add an assert to make this clearer:
```cpp
case ImplicitContextKind::PerThread:
  ...
case ImplicitContextKind::Shared:
 ...
default:
   assert(0); // never happens
```

Reply:
The caller OutgoingAsync checks for null:

https://github.com/zeroc-ice/ice/blob/bf991f4f7604869517f9211f18a6048239521551/cpp/src/Ice/OutgoingAsync.cpp#L696-L705

The null return is documented here:

https://github.com/zeroc-ice/ice/blob/bf991f4f7604869517f9211f18a6048239521551/cpp/include/Ice/Communicator.h#L231-L236